compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -229,7 +239,14 @@ fn layout_of<'tcx>(
         ty::tls::enter_context(&icx, |_| {
             let param_env = param_env.with_reveal_all_normalized(tcx);
             let unnormalized_ty = ty;
-            let ty = tcx.normalize_erasing_regions(param_env, ty);
+
+            let ty = match tcx.try_normalize_erasing_regions(param_env, ty) {
```

Comment:
I think this deserves a comment on why failing to fully normalize is valid here. I would potentially imagine that by the time we need to know layout, we should have all the information to normalize (though I'm not sure, this might not be true)

Reply:
I think it might make sense to have two different versions of `layout_of`. Currently most of the calls of that function happen after typecheck and normalization is always guaranteed to succeed, but there are some calls in typecheck and some lints that use it. Does it makes sense to have one version which requires normalization to succeed and another in which it can fail?

Reply:
This is a good question! In #82039, IIRC we were basically coming to the conclusion that we shouldn't be trying to get the layout of things until after typechecking. So, I think splitting out `layout_of` into two version is a decent intermediate step to "audit", in some way, incorrect behavior.

I think this would be better as a followup. Can you add a FIXME or file an issue?